--- fs/proc/base.c	2026-04-25 13:38:27.195441087 +0530
+++ fs/proc/base.c	2026-04-25 13:40:23.087017185 +0530
@@ -102,6 +102,9 @@
 #include <linux/cn_proc.h>
 #include <linux/task_integrity.h>
 #include <linux/proca.h>
+#if defined(CONFIG_KSU_SUSFS_SUS_MAP) || defined(CONFIG_KSU_SUSFS_OPEN_REDIRECT)
+#include <linux/susfs_def.h>
+#endif
 #include <trace/events/oom.h>
 #include "internal.h"
 #include "fd.h"
